Issue364 input values

dhblum opened the pull request:
This is for #364.  In particular, it:
1. Updates the ``parser.py`` to add an output for every overwrite block.  Consider an overwrite block ``ove1``, then an output is added ``ove1_y`` that is equal to the value of ``ove1.y``.
2. For ``bestest_air``, ``multizone_residential_hydronic``, ``testcase1,2,3``, removes duplicate read blocks in Modelica model and updates unit testing results and documentation for new measurement signals.
3. Updates other unit tests as needed

Still to do:
- [x] Repeat 2. for ``bestest_hydronic`` and ``bestest_hydronic_heat_pump`` test cases.
- [x] Make PR for IDEAS.  
- [x] Merge PR for IDEAS.
- [x] Switch IDEAS back from fork branch
- [x] Reviews
- [x] An update is also now needed with the singlezone_commercial_hydronic test case since that has been merged.  
- [x] Commit all new wrapped.fmus.
- [x] Update release notes, and state this change is non-backward compatible since measurement names change for test cases.
- [x] Update design document as needed.
- [x] Resolve conflicts.

JavierArroyoBastida reviewed (commented):
thanks a lot, @dhblum for this implementation. So far it's looking good to me and It's nice to see that reading overwrite values can be done with such a clean edit in the parser. 
What do you think of establishing a naming convention from now already to distinguish between those outputs that are "current values" of inputs from the other sensor outputs? 
Right now a user can deduce which outputs correspond to overwrite values by checking which are the outputs that have a corresponding input. However, I think it would be clearer if the user could directly identify those inputs from the naming convention. Maybe using a different suffix? something like `_uy` or `_ove_y`? I think it would also help to separate these from the rest of the outputs in the "Model IO's" section of the documentation. I mean, I would add a new subsection to have something like: "Inputs", "Inputs' readings", and "Outputs".

I would also suggest adding some of the new output variables like `con_oveTSetCoo_y` to the list of time series variables to be compared in `testing/compare_references` for visual inspection of changes in the references.

dhblum commented:
@JavierArroyoBastida Thanks for the review.  I've been thinking about this issue of naming convention a bit more. 

Currently, if you call ``/results`` with ``point_name`` equal to one of the control inputs, you get 0s if the control input (_u) is not written (the default value for fmu inputs not given values in pyfmi), otherwise the value specified from the test controller, regardless of whether _activate is 1 or 0.

What if, we simply report in ``/results`` the "current value" of the specified control input (_u).  In this way, the returned values of the control input would equal the baseline control if _activate is 0 and the test controller written value if _activate is 1.  It is our desired performance and much less confusing than the existing way we return a query for control input values described above.  So, within BOPTEST, we still create the FMU output variable ``ove_y`` for every overwrite block (or somehow otherwise mark the ``ove.y`` variable path that needs to be read) but just return its value(s) when ``ove_u`` is the point name requested in ``/results``.  We can also add this to the current state returned from ``/advance``.

dhblum commented on testcases/bestest_air/doc/index.html:
```diff
@@ -279,110 +279,106 @@ <h4>Outputs</h4>
 <ul>
 <li>
-<code>con_reaTSetCoo_y</code> [K] [min=None, max=None]: Zone air temperature setpoint for cooling
+<code>con_oveTSetCoo_y</code> [K] [min=None, max=None]: Zone temperature setpoint for cooling
```
Will make the change.

JavierArroyoBastida commented:
@dhblum what you describe is indeed the desired performance: getting the values of the baseline control if _activate is 0 and the test controller value otherwise. So yes, I agree to return the "current value" of the specified control input, and also to add this to the current state returned from `/advance`. Just for my clarification: according to your proposed implementation, would `ove_y` and `ove_u` always have the same values? or would they differ when the test controller is not overwriting the signal?

dhblum commented:
@JavierArroyoBastida The values of ``ove_u`` returned by ``/advance`` and ``/results`` would always equal ``ove_y``.  Maybe for some even more clarification, if a user specifies as part of ``/advance`` that ``data = {ove_u: 15, ove_activate: 0}``, then ``ove_u`` returned by ``/advanced`` and ``/results`` for that time would equal the baseline control and not equal 15, since _activate is 0.

JavierArroyoBastida commented:
@dhblum what you propose makes sense to me. I would conventionally expose only `ove_u` (instead of `ove_y`) as I think that's more intuitive for a user. The only thing that concerns me a bit is that it would not be possible to know from the results when an input signal comes from the baseline controller or from an external signal, but that's probably something that we don't need to know, right?

dhblum commented:
@JavierArroyoBastida Thanks Javi.  For your last question, do you mean "we" as in BOPTEST developers or "we" as in BOPTEST users?  For developers, if we wanted to track that information we can track the values of ``ove_activate`` within BOPTEST and users can also monitor their own use of ``ove_activate``, or otherwise track their own data submitted to the ``/advance`` requests.

JavierArroyoBastida commented:
I guess I meant "we" as in BOPTEST users, and I agree that users should be able to track their own use of the signals that they overwrite and when they overwrite them in case they need to know whether an input comes from their own controller or from the baseline. 
However, I'm having a hard time trying to understand why the current implementation does not already provide the intended behavior. For example, for the supply air temperature setpoint of the bestest air case, the line that assigns a value to the overwrite variable in this branch is:
```
Modelica.Blocks.Interfaces.RealOutput fcu_oveTSup_y(unit="K") = mod.fcu.oveTSup.y "Supply air temperature setpoint";
```
The assigned value  is thus the one after the switch of the overwrite block, and should therefore show either the value of the external controller OR the one from the baseline controller if `_activate=0`. Then, why is `/results` returning 0s if `_activate=0`?

dhblum commented:
@JavierArroyoBastida Yes, when using ``/results``, retrieving _y would give the correct behavior but retrieving _u does not.  My suggestion in   https://github.com/ibpsa/project1-boptest/pull/377#issuecomment-960014413 was that we match _u to _y, and not expose _y for being retrieved with ``/results``.  I have a first implementation I can push.

JavierArroyoBastida commented:
OK, I see. It's clear now. Thanks! Please push and I'll review.

dhblum commented:
@JavierArroyoBastida Implemented in https://github.com/ibpsa/project1-boptest/pull/377/commits/5e22a912656c82ba6a5fafd50d70c8d19b028dd7.  No unit tests updated or anything yet, but please review.  I tested a bit and think the functionality works.

JavierArroyoBastida commented:
This is looking good to me. Thanks a lot @dhblum for the implementation! I'll follow up when tests are updated. Maybe we can do it stepwise and start by updating tests and reference results for a simple case only, e.g. for testcase1, before going into more complex cases. What do you think? We should be particularly careful with this PR since we can expect it to affect most of the reference results. I'll also double-check the new references.

dhblum commented:
Thanks!  Yea, I'll update reference results carefully, double checking for proper performance.  I'll ping when ready again for review.

dhblum commented:
I've updated functionality, testcases, unit tests, and unit test reference results according to the functionality described in https://github.com/ibpsa/project1-boptest/pull/377#issuecomment-960014413 and discussion thereafter.  I have not yet updated anything for ``bestest_hydronic`` nor ``bestest_hydronic_heat_pump``, which needs updates in the IDEAS library.  I suggest verifying the current changes, and then just making needed updates for those test cases afterwards.

@JavierArroyoBastida and @SenHuang19 can you please review?  @kbenne FYI if you want to review also.

dhblum commented:
FYI I've implemented the needed changes for ``bestest_hydronic`` and ``bestest_hydronic_heat_pump`` on my fork of IDEAS and a local branch of this PR.  Would be ready to make a PR against IDEAS and update this branch pending review of what's already been committed and pushed in this PR.

JavierArroyoBastida reviewed (changes requested):
@dhblum I've gone through this PR again and it's looking good to me in general. I see that references match with historical values. I've just suggested two edits regarding docstrings in two methods from `testcase.py`. I'd also suggest to edit [the compare_references.py file](https://github.com/ibpsa/project1-boptest/blob/14030d7afa7c86e53f19fb0a389c5623da514f5d/testing/compare_references.py#L72) so that bestest air case removes (-) and adds (+) the following check points for visual inspection of reference updates. 
```
-'fcu_reaTSup_y'
-'con_reaTSetCoo_y'
-'con_reaTSetHea_y'
+'fcu_oveTSup_u' 
+'fcu_oveFan_u'
```
Could you also send the corresponding PR to IDEAS already for the bestest_hydronic and bestest_hydronic_heat_pump cases?

dhblum commented on testcase.py:
```diff
@@ -247,8 +266,10 @@ def advance(self,u):
+                z = self._get_full_current_state()
 
-                return self.y
+                return z
```
Done in https://github.com/ibpsa/project1-boptest/pull/377/commits/e3f74f7e610d95721caa852d9e5e4799ad5d43ac.

dhblum commented on testcase.py:
```diff
@@ -302,7 +323,10 @@ def initialize(self, start_time, warmup_period, end_time=np.inf):
+            # Get full current state
+            z = self._get_full_current_state()
+
+            return z
```
Done in https://github.com/ibpsa/project1-boptest/pull/377/commits/e3f74f7e610d95721caa852d9e5e4799ad5d43ac.

dhblum commented:
Thanks @JavierArroyoBastida for the review.  I've updated the compare_references.py in https://github.com/ibpsa/project1-boptest/pull/377/commits/7ba7a6d569d6a038affe98f04f4dc11ca49394d8, and resolved the two other suggestions you made.  I'll prep the PR for IDEAS.  Added remaining to dos to initial comment.

dhblum commented:
Note that Python 2.7 pass not passing since wrapped fmus not updated yet.

dhblum commented:
@JavierArroyoBastida @SenHuang19 @kbenne Pending tests passing I think this is ready to merge.  Please let me know if you have further comments.
